### ISSUE TITLE
The inference terminated without any errors

### ISSUE THREAD

**Issue by yejr0229** [wyysf-98/CraftsMan#1]
I run this command "python inference.py --input eval_data --gpu 0" and it just end without any output and error.Here is the detailed command outputs:
WARNING[XFORMERS]: xFormers can't load C++/CUDA extensions. xFormers was built for:
    PyTorch 2.0.1+cu118 with CUDA 1108 (you have 2.3.0)
    Python  3.10.11 (you have 3.10.14)
  Please reinstall xformers (see https://github.com/facebookresearch/xformers#installing-xformers)
  Memory-efficient attention, SwiGLU, sparse and more won't be available.
  Set XFORMERS_MORE_DETAILS=1 for more details
Some weights of ConditionalCLIPModel were not initialized from the model checkpoint at openai/clip-vit-large-patch14 and are newly initialized: ['vision_model.encoder.layers.0.mod_norm1.mlp.1.bias', 'vision_model.encoder.layers.0.mod_norm1.mlp.1.weight', 'vision_model.encoder.layers.0.mod_norm2.mlp.1.bias', 'vision_model.encoder.layers.0.mod_norm2.mlp.1.weight', 'vision_model.encoder.layers.1.mod_norm1.mlp.1.bias', 'vision_model.encoder.layers.1.mod_norm1.mlp.1.weight', 'vision_model.encoder.layers.1.mod_norm2.mlp.1.bias', 'vision_model.encoder.layers.1.mod_norm2.mlp.1.weight', 'vision_model.encoder.layers.10.mod_norm1.mlp.1.bias', 'vision_model.encoder.layers.10.mod_norm1.mlp.1.weight', 'vision_model.encoder.layers.10.mod_norm2.mlp.1.bias', 'vision_model.encoder.layers.10.mod_norm2.mlp.1.weight', 'vision_model.encoder.layers.11.mod_norm1.mlp.1.bias', 'vision_model.encoder.layers.11.mod_norm1.mlp.1.weight', 'vision_model.encoder.layers.11.mod_norm2.mlp.1.bias', 'vision_model.encoder.layers.11.mod_norm2.mlp.1.weight', 'vision_model.encoder.layers.12.mod_norm1.mlp.1.bias', 'vision_model.encoder.layers.12.mod_norm1.mlp.1.weight', 'vision_model.encoder.layers.12.mod_norm2.mlp.1.bias', 'vision_model.encoder.layers.12.mod_norm2.mlp.1.weight', 'vision_model.encoder.layers.13.mod_norm1.mlp.1.bias', 'vision_model.encoder.layers.13.mod_norm1.mlp.1.weight', 'vision_model.encoder.layers.13.mod_norm2.mlp.1.bias', 'vision_model.encoder.layers.13.mod_norm2.mlp.1.weight', 'vision_model.encoder.layers.14.mod_norm1.mlp.1.bias', 'vision_model.encoder.layers.14.mod_norm1.mlp.1.weight', 'vision_model.encoder.layers.14.mod_norm2.mlp.1.bias', 'vision_model.encoder.layers.14.mod_norm2.mlp.1.weight', 'vision_model.encoder.layers.15.mod_norm1.mlp.1.bias', 'vision_model.encoder.layers.15.mod_norm1.mlp.1.weight', 'vision_model.encoder.layers.15.mod_norm2.mlp.1.bias', 'vision_model.encoder.layers.15.mod_norm2.mlp.1.weight', 'vision_model.encoder.layers.16.mod_norm1.mlp.1.bias', 'vision_model.encoder.layers.16.mod_norm1.mlp.1.weight', 'vision_model.encoder.layers.16.mod_norm2.mlp.1.bias', 'vision_model.encoder.layers.16.mod_norm2.mlp.1.weight', 'vision_model.encoder.layers.17.mod_norm1.mlp.1.bias', 'vision_model.encoder.layers.17.mod_norm1.mlp.1.weight', 'vision_model.encoder.layers.17.mod_norm2.mlp.1.bias', 'vision_model.encoder.layers.17.mod_norm2.mlp.1.weight', 'vision_model.encoder.layers.18.mod_norm1.mlp.1.bias', 'vision_model.encoder.layers.18.mod_norm1.mlp.1.weight', 'vision_model.encoder.layers.18.mod_norm2.mlp.1.bias', 'vision_model.encoder.layers.18.mod_norm2.mlp.1.weight', 'vision_model.encoder.layers.19.mod_norm1.mlp.1.bias', 'vision_model.encoder.layers.19.mod_norm1.mlp.1.weight', 'vision_model.encoder.layers.19.mod_norm2.mlp.1.bias', 'vision_model.encoder.layers.19.mod_norm2.mlp.1.weight', 'vision_model.encoder.layers.2.mod_norm1.mlp.1.bias', 'vision_model.encoder.layers.2.mod_norm1.mlp.1.weight', 'vision_model.encoder.layers.2.mod_norm2.mlp.1.bias', 'vision_model.encoder.layers.2.mod_norm2.mlp.1.weight', 'vision_model.encoder.layers.20.mod_norm1.mlp.1.bias', 'vision_model.encoder.layers.20.mod_norm1.mlp.1.weight', 'vision_model.encoder.layers.20.mod_norm2.mlp.1.bias', 'vision_model.encoder.layers.20.mod_norm2.mlp.1.weight', 'vision_model.encoder.layers.21.mod_norm1.mlp.1.bias', 'vision_model.encoder.layers.21.mod_norm1.mlp.1.weight', 'vision_model.encoder.layers.21.mod_norm2.mlp.1.bias', 'vision_model.encoder.layers.21.mod_norm2.mlp.1.weight', 'vision_model.encoder.layers.22.mod_norm1.mlp.1.bias', 'vision_model.encoder.layers.22.mod_norm1.mlp.1.weight', 'vision_model.encoder.layers.22.mod_norm2.mlp.1.bias', 'vision_model.encoder.layers.22.mod_norm2.mlp.1.weight', 'vision_model.encoder.layers.23.mod_norm1.mlp.1.bias', 'vision_model.encoder.layers.23.mod_norm1.mlp.1.weight', 'vision_model.encoder.layers.23.mod_norm2.mlp.1.bias', 'vision_model.encoder.layers.23.mod_norm2.mlp.1.weight', 'vision_model.encoder.layers.3.mod_norm1.mlp.1.bias', 'vision_model.encoder.layers.3.mod_norm1.mlp.1.weight', 'vision_model.encoder.layers.3.mod_norm2.mlp.1.bias', 'vision_model.encoder.layers.3.mod_norm2.mlp.1.weight', 'vision_model.encoder.layers.4.mod_norm1.mlp.1.bias', 'vision_model.encoder.layers.4.mod_norm1.mlp.1.weight', 'vision_model.encoder.layers.4.mod_norm2.mlp.1.bias', 'vision_model.encoder.layers.4.mod_norm2.mlp.1.weight', 'vision_model.encoder.layers.5.mod_norm1.mlp.1.bias', 'vision_model.encoder.layers.5.mod_norm1.mlp.1.weight', 'vision_model.encoder.layers.5.mod_norm2.mlp.1.bias', 'vision_model.encoder.layers.5.mod_norm2.mlp.1.weight', 'vision_model.encoder.layers.6.mod_norm1.mlp.1.bias', 'vision_model.encoder.layers.6.mod_norm1.mlp.1.weight', 'vision_model.encoder.layers.6.mod_norm2.mlp.1.bias', 'vision_model.encoder.layers.6.mod_norm2.mlp.1.weight', 'vision_model.encoder.layers.7.mod_norm1.mlp.1.bias', 'vision_model.encoder.layers.7.mod_norm1.mlp.1.weight', 'vision_model.encoder.layers.7.mod_norm2.mlp.1.bias', 'vision_model.encoder.layers.7.mod_norm2.mlp.1.weight', 'vision_model.encoder.layers.8.mod_norm1.mlp.1.bias', 'vision_model.encoder.layers.8.mod_norm1.mlp.1.weight', 'vision_model.encoder.layers.8.mod_norm2.mlp.1.bias', 'vision_model.encoder.layers.8.mod_norm2.mlp.1.weight', 'vision_model.encoder.layers.9.mod_norm1.mlp.1.bias', 'vision_model.encoder.layers.9.mod_norm1.mlp.1.weight', 'vision_model.encoder.layers.9.mod_norm2.mlp.1.bias', 'vision_model.encoder.layers.9.mod_norm2.mlp.1.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Restoring states from the checkpoint path at ckpts/image-to-shape-diffusion/clip-mvrgb-modln-l256-e64-ne8-nd16-nl6/model.ckpt with config {'name': 'image-to-shape-diffusion/clip-mvrgb-modln-l256-e64-ne8-nd16-nl6-big', 'description': '', 'tag': 'michelangelo-autoencoder+n4096+noise0.0+pfeat3+normembFalse+lr5e-05+qkvbiasFalse+nfreq8+ln_postTrue', 'seed': 0, 'use_timestamp': True, 'timestamp': '', 'exp_root_dir': 'outputs', 'exp_dir': 'outputs/image-to-shape-diffusion/clip-mvrgb-modln-l256-e64-ne8-nd16-nl6-big', 'trial_name': 'michelangelo-autoencoder+n4096+noise0.0+pfeat3+normembFalse+lr5e-05+qkvbiasFalse+nfreq8+ln_postTrue', 'trial_dir': 'outputs/image-to-shape-diffusion/clip-mvrgb-modln-l256-e64-ne8-nd16-nl6-big/michelangelo-autoencoder+n4096+noise0.0+pfeat3+normembFalse+lr5e-05+qkvbiasFalse+nfreq8+ln_postTrue', 'n_gpus': 8, 'resume': None, 'data_type': 'objaverse-datamodule', 'data': {'root_dir': 'data/objaverse_clean/cap3d_high_quality_170k_images', 'data_type': 'occupancy', 'n_samples': 4096, 'noise_sigma': 0.0, 'load_supervision': False, 'supervision_type': 'occupancy', 'n_supervision': 4096, 'load_image': True, 'image_data_path': 'data/objaverse_clean/raw_data/images/cap3d_high_quality_170k', 'image_type': 'mvrgb', 'idx': [0, 4, 8, 12, 16], 'n_views': 4, 'load_caption': False, 'rotate_points': False, 'batch_size': 32, 'num_workers': 16}, 'system_type': 'shape-diffusion-system', 'system': {'val_samples_json': 'val_data/mv_images/val_samples_rgb_mvimage.json', 'z_scale_factor': 1.0, 'guidance_scale': 7.5, 'num_inference_steps': 50, 'eta': 0.0, 'shape_model_type': 'michelangelo-autoencoder', 'shape_model': {'num_latents': 256, 'embed_dim': 64, 'point_feats': 3, 'out_dim': 1, 'num_freqs': 8, 'include_pi': False, 'heads': 12, 'width': 768, 'num_encoder_layers': 8, 'num_decoder_layers': 16, 'use_ln_post': True, 'init_scale': 0.25, 'qkv_bias': False, 'use_flash': True, 'use_checkpoint': True}, 'condition_model_type': 'clip-embedder', 'condition_model': {'pretrained_model_name_or_path': 'openai/clip-vit-large-patch14', 'encode_camera': True, 'camera_embeds_dim': 32, 'n_views': 4, 'empty_embeds_ratio': 0.1, 'normalize_embeds': False, 'zero_uncond_embeds': True}, 'denoiser_model_type': 'simple-denoiser', 'denoiser_model': {'input_channels': 64, 'output_channels': 64, 'n_ctx': 256, 'width': 768, 'layers': 6, 'heads': 12, 'context_dim': 1024, 'init_scale': 1.0, 'skip_ln': True, 'use_checkpoint': True}, 'noise_scheduler_type': 'diffusers.schedulers.DDPMScheduler', 'noise_scheduler': {'num_train_timesteps': 1000, 'beta_start': 0.00085, 'beta_end': 0.012, 'beta_schedule': 'scaled_linear', 'variance_type': 'fixed_small', 'clip_sample': False}, 'denoise_scheduler_type': 'diffusers.schedulers.DDIMScheduler', 'denoise_scheduler': {'num_train_timesteps': 1000, 'beta_start': 0.00085, 'beta_end': 0.012, 'beta_schedule': 'scaled_linear', 'clip_sample': False, 'set_alpha_to_one': False, 'steps_offset': 1}, 'loggers': {'wandb': {'enable': False, 'project': 'CraftsMan', 'name': 'image-to-shape-diffusion+image-to-shape-diffusion/clip-mvrgb-modln-l256-e64-ne8-nd16-nl6-big+michelangelo-autoencoder+n4096+noise0.0+pfeat3+normembFalse+lr5e-05+qkvbiasFalse+nfreq8+ln_postTrue'}}, 'loss': {'loss_type': 'mse', 'lambda_diffusion': 1.0}, 'optimizer': {'name': 'AdamW', 'args': {'lr': 5e-05, 'betas': [0.9, 0.99], 'eps': 1e-06}}, 'scheduler': {'name': 'SequentialLR', 'interval': 'step', 'schedulers': [{'name': 'LinearLR', 'interval': 'step', 'args': {'start_factor': 1e-06, 'end_factor': 1.0, 'total_iters': 5000}}, {'name': 'CosineAnnealingLR', 'interval': 'step', 'args': {'T_max': 5000, 'eta_min': 0.0}}], 'milestones': [5000]}}, 'trainer': {'num_nodes': 4, 'max_epochs': 100000, 'log_every_n_steps': 5, 'num_sanity_val_steps': 1, 'check_val_every_n_epoch': 3, 'enable_progress_bar': True, 'precision': '16-mixed', 'strategy': 'ddp_find_unused_parameters_true'}, 'checkpoint': {'save_last': True, 'save_top_k': -1, 'every_n_train_steps': 5000}}

**Comment by StriveZs** [comment]
In my sight, the inference code is not complete, like:
`model = load_model(ckpt_path, config_path, device)
    
    # read the input images
    if os.path.isdir(args.input):
        image_files = glob(os.path.join(args.input, "*.png"))
    else:
        image_files = [args.input]
    
    for image_file in image_files:
        pass
    
    rmbg = RMBG(device)
    gen_mvimg = GenMVImage(device)

    for image_file in image_files:
        image = Image.open(image_file)
        import pdb; pdb.set_trace()
        # image = rmbg.rmbg_sam(image, foreground_ratio=0.5)
        # mv_img = gen_mvimg.gen_image_from_crm(image)
        # for i, img in enumerate(mv_img):
        #     img.save(f"{args.output}/mv_img_{i}.png")
        # print(f"Multi-view images saved to {args.output}")`

The model has not been used anywhere.

Moreover, I have carefully read the training code in base.py, which is also not complete, it includes so many "raise NotImplementedError".

**Comment by yejr0229** [comment]
I see,so I'd like to know when will you complete it?

**Comment by wyysf-98** [comment]
Hi, I will finish this part today. :)

**Comment by StriveZs** [comment]
Thx

> Hi, I will finish this part today. :)

**Comment by yejr0229** [comment]
> Hi, I will finish this part today. :)

That's so great

**Comment by wyysf-98** [comment]
> In my sight, the inference code is not complete now, like:
> 
> ```
> model = load_model(ckpt_path, config_path, device)  
> # read the input images
> if os.path.isdir(args.input):
>     image_files = glob(os.path.join(args.input, "*.png"))
> else:
>     image_files = [args.input]
> 
> for image_file in image_files:
>     pass
> 
> rmbg = RMBG(device)
> gen_mvimg = GenMVImage(device)
> 
> for image_file in image_files:
>     image = Image.open(image_file)
>     import pdb; pdb.set_trace()
>     # image = rmbg.rmbg_sam(image, foreground_ratio=0.5)
>     # mv_img = gen_mvimg.gen_image_from_crm(image)
>     # for i, img in enumerate(mv_img):
>     #     img.save(f"{args.output}/mv_img_{i}.png")
>     # print(f"Multi-view images saved to {args.output}")
> ```
> 
> The model has not been used anywhere.
> 
> Moreover, I have carefully read the training code in base.py, which is also not complete, it includes so many "raise NotImplementedError".

hello, the training code is complete, the 'NotImplementedError' is just for checking the correct config. I have tested the training code and the inference code is TBD as in README.md as we released the gradio demo first.

**Comment by yejr0229** [comment]
If convenient, may I ask when will the inference code completed?

**Comment by wyysf-98** [comment]
> If convenient, may I ask when will the inference code completed?

Hi, I'm almost done and will update after checking. BTW, you can try our hugging face demo first: https://huggingface.co/spaces/wyysf/CraftsMan

**Comment by StriveZs** [comment]
Hi, I want to confirm some mistakes in the project code with you the class of MichelangeloAutoencoder in the /CraftsMam/craftsman/models/michelangelo_autoencoder.py has missed the forward function？ Today I read your code following the training logic and I found the above question. Could you please give me some suggestions or solve my confusion?

**Comment by wyysf-98** [comment]
> Hi, I want to confirm some mistakes in the project code with you the class of MichelangeloAutoencoder in the /CraftsMam/craftsman/models/michelangelo_autoencoder.py has missed the forward function？ Today I read your code following the training logic and I found the above question. Could you please give me some suggestions or solve my confusion? @wyysf-98

We abstracted the `autoencoder` class and the MichelangeloAutoencoder inherits from the `autoencoder` class, so the forward function is: https://github.com/wyysf-98/CraftsMan/blob/main/craftsman/models/autoencoders/utils.py#L192
